### PR TITLE
[vsp] Add VspHost and Pubkey to udb

### PR DESCRIPTION
**Pull request by alexlyp**
This PR adds 2 new buckets to the database for storing vsp host and pubkey information 
for vsp tickets.   The VSP host bucket is keyed by a basic index (that is referenced in the VSPTicket
entries and the VSP Pubkey bucket is keyed by hostname. 

Together these allow us to provide VSP Host and Pubkey information for each ticket.

There is now also an additional VSPFeeStatus for Confirmed.  This allows us to skip any previously
confirmed ticket from being checked again.

### CODE REVIEW

**Comment by jholdstock** [comment]
Sounds like this would close #2059

**Comment by alexlyp** [comment]
Yah I believe that's accurate @jholdstock.  Hopefully this will catch all confirmed tickets once and then skip over them on any subsequent processing.

File: wallet/udb/vsp.go [hunk 1]
```diff
@@ -26,15 +28,55 @@ const (
...
+
+		err = SetVSPPubKey(dbtx, []byte(record.Host), &VSPPubKey{
+			ID:     vspHostID,
+			PubKey: record.PubKey,
+		})
+		if err != nil {
+			return err
+		}
+		record.VSPHostID = vspHostID
+	} else {
```
Comment (matheusd) [inline wallet/udb/vsp.go:76]: You need to add a test for `err != nil` somewhere, otherwise this is also executed when the `GetVSPPubKey` failed for other reasons.

File: wallet/udb/vsp.go [hunk 2]
```diff
@@ -52,6 +94,24 @@ func GetVSPTicket(dbtx walletdb.ReadTx, tickethash chainhash.Hash) (*VSPTicket,
 	}
 	ticket := deserializeVSPTicket(serializedTicket)
 
+	host, err := GetVSPHost(dbtx, ticket.VSPHostID)
+	if err != nil && !errors.Is(err, errors.NotExist) {
+		// Only error out if it's not a 'not exist error'
+		return nil, err
+	}
```
Comment (matheusd) [inline wallet/udb/vsp.go:101]: Same as above, you're ignoring any errors that are not `errors.NotExist`

File: wallet/udb/vsp.go [hunk 3]
```diff
@@ -52,6 +94,24 @@ func GetVSPTicket(dbtx walletdb.ReadTx, tickethash chainhash.Hash) (*VSPTicket,
...
+	if host != nil && len(host.Host) > 0 {
+		// If the stored host is not empty then get the saved pubkey as well.
+		ticket.Host = string(host.Host)
+		pubkey, err := GetVSPPubKey(dbtx, host.Host)
+		if err != nil && !errors.Is(err, errors.NotExist) {
+			// XXX should we error here when pubkey isn't found but host is?
```
Comment (matheusd) [inline wallet/udb/vsp.go:107]: I think it's fine as is (i.e. don't error out if there is a host without an associated pubkey) but we might want to add a sanity check earlier in the ticket buying stack to ensure we have both host and valid pubkey before starting a vsp ticket purchase.

File: wallet/udb/vsp.go [hunk 4]
```diff
@@ -26,15 +28,55 @@ const (
...
+	VSPHostID   uint32
+	Host        string
+	PubKey      []byte
 }
 
 // SetVSPTicket sets a VSPTicket record into the db.
 func SetVSPTicket(dbtx walletdb.ReadWriteTx, ticketHash *chainhash.Hash, record *VSPTicket) error {
+	// Check for Host/Pubkey buckets
+	pubkey, err := GetVSPPubKey(dbtx, []byte(record.Host))
+	if err != nil && errors.Is(err, errors.NotExist) {
```
Comment (matheusd) [inline wallet/udb/vsp.go:48]: Re-reding, one limitation here is that this doesn't allow replacing the pubkey for the given host if it ever changes. 

Maybe after fetching the pubkey via `getVSPPubKey`, check if it is `!= nil` and `!= record.PubKey` and then rewrite the pubkey in the DB to use the new one (while keeping the host ID intact). This would future-proof if a VSP ever changes its pubkey for any reason.

File: internal/vsp/vsp.go [hunk 5]
```diff
@@ -138,6 +138,15 @@ func (c *Client) ProcessUnprocessedTickets(ctx context.Context, policy Policy) {
 		if err == nil {
 			return nil
 		}
+		confirmed, err := c.Wallet.IsVSPTicketConfirmed(ctx, hash)
+		if err != nil && !errors.Is(err, errors.NotExist) {
+			// Should we just log and return nil here?
+			return err
```
Comment (matheusd) [inline internal/vsp/vsp.go:144]: Probably log either here or this needs to be refactored so that the error returned in the `c.ForUnspentUnexpiredTickets` above isn't simply ignored.

File: wallet/udb/vsp.go (hunk 2 above)
Comment (alexlyp) [reply wallet/udb/vsp.go:101]: I believe this one is ok since it's !errors.Is(err, errorsNotExist).  So it should only be erroring on errors that aren't non exist errors.

File: wallet/udb/vsp.go (hunk 3 above)
Comment (alexlyp) [reply wallet/udb/vsp.go:107]: Good point, I'll add that sanity check.

Comment (alexlyp) [reply wallet/udb/vsp.go:107]: After taking a more in depth look, I believe it's currently impossible to submit a purchase ticket (CLI, grpc with puchase ticket and autobuyer).  All error if vsp host is provided and not pubkey (or vice versa).  So I think we're safe here.

File: wallet/udb/vsp.go [hunk 6]
```diff
@@ -73,9 +75,20 @@ func SetVSPTicket(dbtx walletdb.ReadWriteTx, ticketHash *chainhash.Hash, record
...
-	} else {
-		record.VSPHostID = pubkey.ID
+	} else if err != nil {
+		return err
+	}
+
+	// If the pubkey from the record in the request differs from the pubkey
+	// in the database that is saved for the host, update the pubkey in the
+	// db but keep the vsphost id intact.
+	if pubkey.PubKey != nil && bytes.Compare(pubkey.PubKey, record.PubKey) != 0 {
```
Comment (matheusd) [inline wallet/udb/vsp.go:85]: `GetVSPPubKey` will return a `nil` `pubkey` if it returns `errors.NotExist` so this will panic in that case. You need to test whether `pubkey` is nil (not `pubkey.PubKey`). `bytes.Compare` can handle a nil `pubkey.PubKey` (though that won't happen in practice).

File: wallet/udb/vsp.go [hunk 7]
```diff
@@ -73,9 +75,20 @@ func SetVSPTicket(dbtx walletdb.ReadWriteTx, ticketHash *chainhash.Hash, record
...
+		err = SetVSPPubKey(dbtx, []byte(record.Host), &VSPPubKey{
+			ID:     pubkey.ID,
+			PubKey: record.PubKey,
+		})
 	}
+	record.VSPHostID = pubkey.ID
```
Comment (matheusd) [inline wallet/udb/vsp.go:91]: Also, this needs to be inside the above `if pubkey != nil ...`.

**Comment by matheusd** [comment]
This tested fine after locally implementing the changes I mentioned in the review (`ticketinfo` command returned the vsp info).

File: wallet/udb/vsp.go (hunk 6 above)
Comment (alexlyp) [reply wallet/udb/vsp.go:85]: ah good catch

**Comment by alexlyp** [comment]
Ok fixed your noted issues @matheusd.  I've also added a request from @jrick about checking db status before continuing processing.  Now we check that the fee is errored or started before attempting to pay/send again.  If it's already list as paid, then only confim the payment (but must be at the same vsp used previously).  And if it is listed as confirmed, do nothing.

File: wallet/wallet.go [hunk 8]
```diff
@@ -5520,6 +5520,23 @@ func (w *Wallet) SetPublished(ctx context.Context, hash *chainhash.Hash, publish
 	return nil
 }
 
+// VSPTicketInfo returns the various information for a given vsp ticket
+func (w *Wallet) VSPTicketInfo(ctx context.Context, ticketHash *chainhash.Hash) (*udb.VSPTicket, error) {
+	data := &udb.VSPTicket{}
```
Comment (matheusd) [inline wallet/wallet.go:5525]: This should be declared as `var data *udb.VSPTicket` to avoid allocating a value that is ultimately discarded (due to the next `data, err = udb.GetVSPTicket(..` always replacing this value.

Also, I notice this is returning a struct from the `udb` package, which I know @jrick wants to ultimately make an internal package, so maybe this is going to need to export an adapted struct instead (he can probably chime in on whether this should be done in this PR or left for a future work).

File: internal/vsp/vsp.go [hunk 9]
```diff
@@ -238,39 +239,69 @@ func (c *Client) Process(ctx context.Context, ticketHash *chainhash.Hash, feeTx
...
+	if err != nil && !errors.Is(err, errors.NotExist) {
+		log.Error(err)
+		return nil
 	}
-	fp.mu.Unlock()
-	err := fp.receiveFeeAddress()
-	if err != nil {
-		err := c.Wallet.UpdateVspTicketFeeToErrored(ctx, ticketHash, c.client.url, c.client.pub)
+
+	switch udb.FeeStatus(vspTicket.FeeTxStatus) {
```
Comment (matheusd) [inline internal/vsp/vsp.go:249]: This needs to also check first whether `vspTicket` is `!= nil`, otherwise this may panic if `ProcessWithPolicy` is ever somehow called with a tx hash that is not a previously added ticket. 

Either that, or change the previous if to _not_ ignore `errors.NotExist`.

File: wallet/wallet.go (hunk 8 above)
Comment (alexlyp) [reply wallet/wallet.go]: Yah good point about the struct.  Added to the wallet package, maybe it makes more sense somewhere if you guys would like.

File: internal/vsp/vsp.go [hunk 10]
```diff
@@ -242,14 +242,12 @@ func (c *Client) ProcessWithPolicy(ctx context.Context, ticketHash *chainhash.Ha
...
+		return err
 	}
 
-	switch udb.FeeStatus(vspTicket.FeeTxStatus) {
-	// If VSPTicket has been started or errored then attempt to create a new fee
-	// transaction, submit it then confirm.
-	case udb.VSPFeeProcessStarted, udb.VSPFeeProcessErrored:
+	// VSPTicket info doesn't exists yet so we need to do all the basic
+	// processing.
+	if vspTicket == nil {
```
Comment (matheusd) [inline internal/vsp/vsp.go:250]: There's now a chunk of non-trivial code that is (and will always need to be) duplicated across the `vspTicket == nil` and `case udb.VSPFeeProcessStarted, udb.VSPFeeProcessErrored` cases, which (while correct now) could trigger a bug in the future if one of the instances is modified by the other is not.

I suggest simplifying this to:

```go
feeStatus := udb.VSPFeeProcessStarted // Will be used if the ticket isn't registered to the vsp yet.
if vspTicket != nil {
  feeStatus = vspTicket.FeeTxStatus
}

switch (feeStatus) {
case udb.VSPFeeProcessStarted, udb.VSPFeeProcessErrored:
  // ...
// other cases
}
```

**Review by jrick — APPROVED** [review]
